### PR TITLE
Fix legend.justification

### LINKED ISSUE #1716: Control justification of legends
I'm not sure why there's currently no way to do this. Is this what `legend.box.just` is supposed to control?

I think the default would look nicer if the legends were left/top aligned to the panel area.

### PR DESCRIPTION
**Pull request by thomasp85**
This PR fixes #1716 by stretching the viewport reserved for legend drawing.

### CODE REVIEW

**Comment by hadley** [comment]
That's all that's needed???

Can you show a couple of examples?

**Comment by thomasp85** [comment]
I'm that good :-)

... was what I thought I would say until I tested it out - the alignment is related to the centre of the legend gtable (i.e. the center is flush with the top of the panel). I'll find a way to fix this and then add some examples as well

**Comment by thomasp85** [comment]
Hmm... we might want a `legend.anchor` that determines where the justification is measured from...

**Comment by thomasp85** [comment]
This works (though I have a hard time understanding why it works for all justification values 😖) <- thats apparently a confused smiley


Examples:

```r
p <- ggplot(mtcars) + geom_point(aes(x = cyl, y = mpg, colour = gear, size = hp))

p
```
rplot01

```r
p + theme(legend.justification = 'top')
```
rplot02

```r
p + theme(legend.justification = 'bottom')
```
rplot03

```r
p + theme(legend.justification = 'left', legend.position = 'bottom', legend.box.just = 'left')
```
rplot04

**Comment by hadley** [comment]
Nice!  

What do you think about making top/left alignment the default?

File: R/plot-build.r [hunk 1]
```diff
@@ -215,7 +215,8 @@ ggplot_gtable <- function(data) {
     } else {
       # x and y are adjusted using justification of legend box (i.e., theme$legend.justification)
       legend_box <- editGrob(legend_box,
-        vp = viewport(x = xjust, y = yjust, just = c(xjust, yjust)))
+        vp = viewport(x = xjust, y = yjust, just = c(xjust, yjust),
+          height = unit(1, "null")+legend_height, width = unit(1, "null")+legend_width))
```
Comment (hadley) [inline R/plot-build.r]: Needs spaces around `+`

**Comment by hadley** [comment]
Can you please add a bullet to NEWS, and update the theme docs to indicate that it also adjusts the justification outside of the plot?

(Oh and maybe check it doesn't break when `legend.position` is a two-element numeric vector)

**Comment by thomasp85** [comment]
The current "fad" is to align everything against the borders of the plotting area so this would fall in line with that. Whether it is more visually pleasing etc. is beyond me but it would result in less "rearrangement" when additional guides are added...

What are your thoughts on this vs approach two?

**Comment by hadley** [comment]
I think it's more aesthetically pleasing to align to borders (I can try and explain why when we talk), but it would also require some tweaking so that the legend padding only occurs between legends, not on the end.

I don't remember enough to know which is better, so I'd be inclined to go with @baptiste's advice.

**Comment by baptiste** [comment]
BTW, could all `theme()` arguments be explicitly listed in the function's signature? It would be immensely helpful to have tab completion...

**Comment by hadley** [comment]
@baptiste I'd accept a PR that implemented that (you'd still need ... as a catch all)

**Comment by thomasp85** [comment]
I personally think that version 2 is cognitively easier to reason about (though the results are equivalent) at the expense of creating a slightly more complex gtable.

With that version we can definitely say that intermediary justification (e.g. 0.3) controls the relative size of the top and bottom margin...

**Comment by thomasp85** [comment]
Both versions works fine with numeric justification btw

**Comment by thomasp85** [comment]
As I discus in #1716 I think we should separate the margins around the complete legend area from the margins around each guide so it is possible to get the legend completely flush with the top without needing to give up spacing between guides...

It's a quick fix - should I make a separate PR or lump it in together with either this or #1725?

`legend.box.margin` would be the natural name I think, but I do believe that `legend.box.just` and `legend.justification` in a perfect world should be switched in meaning...

**Comment by hadley** [comment]
Let's go with the other approach. And I think you might as well incorporate the margin stuff, which now includes #1727

**Pull request closed without merging**